Fix #192 - Cannot open recording stream on Moto G

dturner opened the pull request.

philburk commented on src/opensles/AudioInputStreamOpenSLES.cpp:
```diff
@@ -124,11 +126,29 @@ Result AudioInputStreamOpenSLES::open() {
+    if (result == SL_RESULT_PARAMETER_INVALID && usingExtendedDataFormat){
+
+        LOGW("Unable to open recording stream using PCM_EX format, trying again using PCM...");
+        audioSink.pFormat = &format_pcm;
```
Should we set usingExtendedDataFormat=false here? For line 150 below.

dturner replied:
Doh! Thanks Phil - this is exactly why I ask you to review :)

dturner replied:
Fixed now

philburk commented on src/opensles/AudioInputStreamOpenSLES.cpp:
```diff
@@ -124,11 +126,30 @@ Result AudioInputStreamOpenSLES::open() {
 
+    // If the stream is not using the extended data format it must have a format of I16 since no
+    // other format is supported, regardless of what format was requested.
+    if (!usingExtendedDataFormat) mFormat = AudioFormat::I16;
```
In AAudio, if you request Format=UNSPECIFIED then you will get either I16 or FLOAT.
If you request FLOAT then you will get a FLOAT stream or the stream will not open.
You do not get an I16 stream. Same for channels.

So we should probably do something like:
    if (formatResuested == UNSPECIFIED) mFormat = I16
    else return ERROR;

dturner replied:
> or the stream will not open

That's interesting. So it's not like PerformanceMode where if you request LowLatency but don't get it the stream still opens but with PerformanceMode=None.

In that case, like you say, we should make the behaviour the same for OpenSL ES streams. Which is (correct me if I'm wrong)...

For playback streams
==

**API < 21 (only I16 supported)**
Unspecified - Stream opens with format set to I16
Float - Stream does not open and returns relevant error (should we add `ErrorUnsupportedAudioFormat` so it's clear the reason the stream didn't open?)
I16 - Stream opens with format set to I16

**API 21+ (I16 and float supported)**
Unspecified - Stream opens with format set to Float
Float - Stream opens with format set to Float
I16 - Stream opens with format set to I16

For recording streams:
==
**API < 23 (only I16 supported)**
Unspecified - Stream opens with format set to I16
Float - Stream does not open and returns relevant error
I16 - Stream opens with format set to I16

**API 23+ (I16 and float supported)**
Unspecified - Stream opens with format set to Float
Float - Stream opens with format set to Float
I16 - Stream opens with format set to I16

philburk replied:
Right. If a caller request a specific format then they should either get that format or it should fail.

dturner commented:
Since this behaviour is now defined (see comment above) I have refactored `Audio*StreamOpenSLES` and added tests. New tests passing on API 18, 21, 23 and 28.

dturner commented:
@philburk  Any chance of a +1 here?
